src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs

```diff
@@ -2,78 +2,364 @@
+        public static MethodInvoker Create(MethodInfo method)
         {
-            if (_invokeFunc != null && (invokeAttr & BindingFlags.DoNotWrapExceptions) != 0)
+            ArgumentNullException.ThrowIfNull(method, nameof(method));
+
+            if (method is not RuntimeMethodInfo runtimeMethod)
             {
-                return _invokeFunc(obj, args);
+                throw new ArgumentException(SR.Argument_MustBeRuntimeMethodInfo, nameof(method));
             }
-            return Invoke(obj, args, invokeAttr);
+
+            return new MethodInvoker(runtimeMethod);
         }
 
-        private bool _invoked;
-        private bool _strategyDetermined;
-        private InvokerEmitUtil.InvokeFunc? _invokeFunc;
+        private MethodInvoker(MethodBase method, RuntimeType[] argumentTypes)
+        {
+            if (LocalAppContextSwitches.ForceInterpretedInvoke && !LocalAppContextSwitches.ForceEmitInvoke)
+            {
+                // Always use the native invoke; useful for testing.
+                _strategy = InvokerStrategy.StrategyDetermined_Obj4Args | InvokerStrategy.StrategyDetermined_ObjSpanArgs | InvokerStrategy.StrategyDetermined_RefArgs;
+            }
+            else if (LocalAppContextSwitches.ForceEmitInvoke && !LocalAppContextSwitches.ForceInterpretedInvoke)
+            {
+                // Always use emit invoke (if IsDynamicCodeSupported == true); useful for testing.
+                _strategy = InvokerStrategy.HasBeenInvoked_Obj4Args | InvokerStrategy.HasBeenInvoked_ObjSpanArgs | InvokerStrategy.HasBeenInvoked_RefArgs;
+            }
+
+            _method = method;
+            _argTypes = argumentTypes;
+            _argCount = _argTypes.Length;
+
+            MethodInvokerCommon.Initialize(argumentTypes, ref _strategy, out _invokerArgFlags, out _needsByRefStrategy);
+        }
 
-        [DebuggerStepThrough]
-        [DebuggerHidden]
-        private unsafe object? Invoke(object? obj, IntPtr* args, BindingFlags invokeAttr)
+        public object? Invoke(object? obj) => Invoke(obj, null, null, null, null);
+        public object? Invoke(object? obj, object? arg1) => Invoke(obj, arg1, null, null, null);
+        public object? Invoke(object? obj, object? arg1, object? arg2) => Invoke(obj, arg1, arg2, null, null);
+        public object? Invoke(object? obj, object? arg1, object? arg2, object? arg3) => Invoke(obj, arg1, arg2, arg3, null);
+        public object? Invoke(object? obj, object? arg1, object? arg2, object? arg3, object? arg4)
         {
-            if (!_strategyDetermined)
+            if ((_invocationFlags & (InvocationFlags.NoInvoke | InvocationFlags.ContainsStackPointers)) != 0)
+            {
+                ((RuntimeMethodInfo)_method).ThrowNoInvokeException();
+            }
+
+            if (_argCount > MaxStackAllocArgCount)
+            {
+                MethodBaseInvoker.ThrowTargetParameterCountException();
+            }
+
+            if (!_method.IsStatic)
```

Comment:
`MethodBase.IsStatic` is virtual call. This should be cached. Cache the runtime type of the target method and use it to validate `obj` when it is non-null?

Reply:
It is non-virtual wrapper of a virtual call.

Attributes property is abstract/virtual: 
https://github.com/dotnet/runtime/blob/d52b3038cd19aab5b95be4f1c76010b91dd07429/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBase.cs#L18

Reply:
Thanks